lib/maths/CBoostedTreeFactory.cc

```diff
@@ -299,20 +317,58 @@ void CBoostedTreeFactory::initializeUnsetRegularizationHyperparameters(core::CDa
     //   1) Get percentile estimates of the gain in the loss function and its sum
     //      curvature from the splits selected in a single tree with regulizers
     //      zeroed,
-    //   2) Use these to upper bound the size of gamma and lambda, that is find
-    //      values for which we expect to underfit the data,
+    //   2) Use these to upper bound the size of alpha, gamma and lambda, that is
+    //      find values we expect to underfit the data,
     //   3) Decrease each regularizer and look for a turning point in the test
     //      loss, i.e. the point at which transition to overfit occurs.
     // We'll search intervals in the vicinity of these values in the hyperparameter
     // optimisation loop.
 
     core::CPackedBitVector allTrainingRowsMask{m_TreeImpl->allTrainingRowsMask()};
 
+    double log2MaxTreeSize{std::log2(
+        static_cast<double>(m_TreeImpl->maximumTreeSize(allTrainingRowsMask)))};
+    m_TreeImpl->m_Regularization.maxTreeDepth(
+        m_TreeImpl->m_RegularizationOverride.maxTreeDepth().value_or(log2MaxTreeSize));
+    m_TreeImpl->m_Regularization.maxTreeDepthTolerance(
+        m_TreeImpl->m_RegularizationOverride.maxTreeDepthTolerance().value_or(
+            0.5 * (MIN_MAX_DEPTH_TOLERANCE + MAX_MAX_DEPTH_TOLERANCE)));
+    LOG_TRACE(<< "max depth = " << m_TreeImpl->m_Regularization.maxTreeDepth()
+              << ", tol = " << m_TreeImpl->m_Regularization.maxTreeDepthTolerance());
+
     double gainPerNode;
     double totalCurvaturePerNode;
     std::tie(gainPerNode, totalCurvaturePerNode) =
         this->estimateTreeGainAndCurvature(frame, allTrainingRowsMask);
 
+    if (m_TreeImpl->m_RegularizationOverride.alpha() == boost::none) {
+        if (gainPerNode > 0.0) {
+            TVector fallbackInterval{{std::log(MIN_REGULARIZER_SCALE), 0.0,
+                                      std::log(MAX_REGULARIZER_SCALE)}};
+            fallbackInterval += TVector{std::log(m_TreeImpl->m_Eta)};
+
+            double logInitialAlpha{std::log(gainPerNode)};
+            auto applyAlphaStep = [logInitialAlpha](CBoostedTreeImpl& tree,
+                                                    double stepSize, std::size_t step) {
+                tree.m_Regularization.alpha(
+                    std::exp(logInitialAlpha + static_cast<double>(step) * stepSize));
+            };
+            m_LogAlphaSearchInterval =
+                TVector{std::log(gainPerNode)} +
+                this->testLossNewtonLineSearch(frame, allTrainingRowsMask, applyAlphaStep,
```

Comment:
If you are changing the code here anyway, could you please remove "Newton" from the name of the function. During our last conversation, I confused it with a Newton optimization, since there is a step size and steps, :see_no_evil:  but we are dealing with quadratic approximation here and minimization happens elsewhere.

Reply:
I think Newton is ok in the sense it is using a quadratic approximation to find the minimum. I agree it isn't conventional iterated Newton's method, but I still think it captures the gist of what we're doing here. I could switch back to something like `lineSearchForMinimumQuadraticTestLoss` if you prefer.

Reply:
At this point, I would opt for `testLossLineSearch` for simplification.

Reply:
Done in [e732fe7](https://github.com/elastic/ml-cpp/pull/698/commits/e732fe770fb9d4532853c147c655b98a248dec0d)